Fix some incorrect SpecialFolder entries for Unix

Miepee opened the pull request:
This PR:
- fixes an incorrect Documents/Personal entry for Linux+Mac (from `HOME` to `XDG_DOCUMENTS_DIR`/`HOME/Documents` for Linux and Mac respectively)
- fixes an incorrect Videos entry for Mac (from `HOME/Videos` to `HOME/Movies`)
- fixes an incorrect (Local) ApplicationData entry for Mac (from `XDG_CONFIG_HOME` and `XDG_DATA_HOME` to `HOME/Library/Application Support` for both ApplicationData and LocalApplicationData)

- Changes the Unix tests to check that `SpecialFolder.Personal` and `SpecialFolder.MyDocuments` are the Documents directory instead of `HOME`.

Fixes #63214, 
Fixes #68323, 
Fixes #40353

Miepee commented:
Hm, a bit confused on that failed Windows Arm build. Can't see anything I did that broke it, nor see from the logs on where it error'd on my changes.

danmoseley commented:
It's not you that broke these tests. @MattGal  if this from helix?

running %HELIX_CORRELATION_PAYLOAD%\scripts\ce25ee46de8f44688fa71d0af86d03fb\execute.cmd in C:\h\w\BEED09AC\w\A71708E3\e max 900 seconds

Output:

Exit Code:-900

MattGal commented:
> It's not you that broke these tests. @MattGal is this from helix?
> 
> running %HELIX_CORRELATION_PAYLOAD%\scripts\ce25ee46de8f44688fa71d0af86d03fb\execute.cmd in C:\h\w\BEED09AC\w\A71708E3\e max 900 seconds
> 
> Output:
> 
> Exit Code:-900

Yes we hotfix-rolled-back a change that was causing this last week (https://github.com/dotnet/runtime/issues/67728) .  If you're still seeing it today we should investigate but checking out similar runs I think we're OK.  We may need some variant of the fix back though, but based on the date that's all this is.

Miepee commented:
Since [os-mac-os-x](https://github.com/dotnet/runtime/labels/os-mac-os-x) got labeled for this, someone might wanna label this with [os-linux](https://github.com/dotnet/runtime/labels/os-linux) as well.

danmoseley commented:
To enumerate breaking changes, on Mac, AppData goes from `/home/.config` today to `/home/Library/Application Support` and LocalAppData goes from `/home/.local/share` to `/home/Library/` and Documents/Personal from `/home` to `/home/Documents` and Videos from `/home/Videos` to `/home/Movies` . While on Linux, Documents/Personal goes from `/home` to XDG_DOCUMENTS_DIR.

@marklio thoughts about the impact of the break here? As always, it's a major update, so if we think it's the right change, it may well be acceptable to correct all these.

@dotnet/area-system-runtime thoughts?

It's quite remarkable how much discussion this has involved, and this is only some of the proposed changes. https://github.com/dotnet/runtime/issues/63214 is just one of the long discussions.

My inclination is to take this and look for feedback on Preview 5+.

danmoseley commented:
The code changes look fine to me though.

danmoseley commented:
rerunning validation

jkotas commented:
> thoughts about the impact of the break here? 

The mobile platforms are probably the most impacted here.

cc @marek-safar

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:
```diff
@@ -85,31 +85,23 @@ private static string GetFolderPathCoreWithoutValidation(SpecialFolder folder)
-                case SpecialFolder.MyVideos:
-                    return ReadXdgDirectory(home, "XDG_VIDEOS_DIR", "Videos");
-
 #if TARGET_OSX
```
Should this ifdef be enabled for all Apple platforms?

Miepee replied:
Good question. I sadly only have one Apple device which is running MacOS.
Don't the mobile platforms get the folder from a different place though? I thought they'd get it from [here](https://github.com/dotnet/runtime/blob/fa284882f1fec209d2cadd915c621078e8041228/src/mono/System.Private.CoreLib/src/System/Environment.iOS.cs#L35) or is this mono specific?

danmoseley replied:
It does [not apply for iOS](https://github.com/dotnet/runtime/blob/3535e0769f202ae4cd820bea24afd20cee313966/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems#L2176). Is that all other Apple platforms?

Miepee replied:
From quick skimming the RIDs, there's TVOS and MacCatalyst as well. MacCatalyst is inheriting from iOS, tvOS is only inheriting from unix. I cannot however find much documentation regarding tvOS' file system right now. Does seem relatively locked down though, with apps not meant to write to local storage and [mostly to the cloud instead](https://developer.apple.com/library/archive/documentation/General/Conceptual/AppleTV_PG/iCloudStorage.html#//apple_ref/doc/uid/TP40015241-CH10-SW1).

jkotas replied:
> It does not apply for iOS

Ah ok, I have missed that this file is ifdefed out.

Is there a good reason for the default folder handling to differ between macOS and other Apple OSes? Should macOS be switched to just use https://github.com/mono/mono/blob/main/mcs/class/corlib/System/Environment.iOS.cs?

Miepee replied:
Am not that familiar with the codebase, but my two cents are, that it probably would be better to put macOS into there as well, rename the file to `Environment.Apple`, and put some macOS specific checks in there (for i.e. `ApplicationData`, `Personal` etc.).

On another note, why does `Environment.iOS` hardcode the subdirectories, instead of invoking the respective NSLibrary paths?
For `SpecialFolder.Desktop` for example, `NSDocumentDirectory/Desktop` is used instead of `NSDesktopDirectory` which according to the Apple docs at least work on [all Apple OS](https://developer.apple.com/documentation/foundation/nssearchpathdirectory/nsdesktopdirectory). This is probably a thing for a separate PR, but was curious nonetheless.

(Also, [updated link](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Environment.iOS.cs))

rolfbjarne replied:
I agree that if you're going to do breaking changes, do it right by using [NSSearchPathDirectory](https://developer.apple.com/documentation/foundation/nssearchpathdirectory) values as much as possible, like it's done for iOS here:

https://github.com/dotnet/runtime/blob/16cb35fda05165d679603c1c9c7ca5710e9bd3df/src/libraries/System.Private.CoreLib/src/System/Environment.iOS.cs#L84-L85

Miepee replied:
As far as I can see, that'd require changing the method's return value from `string` to `string?` (as the `Interop.Sys.SearchPath` method can return null), which would be yet another API break. Is that a change worth having?

danmoseley commented:
What Mono had, for comparison [here](https://github.com/mono/mono/blob/main/mcs/class/corlib/System/Environment.cs) and for iOS [here](https://github.com/mono/mono/blob/main/mcs/class/corlib/System/Environment.iOS.cs) 

We have code specific for iOS [here](https://github.com/dotnet/runtime/blob/3535e0769f202ae4cd820bea24afd20cee313966/src/mono/System.Private.CoreLib/src/System/Environment.iOS.cs#L35) and for Android [here](https://github.com/dotnet/runtime/blob/3535e0769f202ae4cd820bea24afd20cee313966/src/libraries/System.Private.CoreLib/src/System/Environment.Android.cs#L16)

tmds commented on src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:
```diff
@@ -85,31 +85,23 @@ private static string GetFolderPathCoreWithoutValidation(SpecialFolder folder)
+                case SpecialFolder.ApplicationData:
+                    return Path.Combine(home, "Library", "Application Support");
+                case SpecialFolder.LocalApplicationData:
+                    return Path.Combine(home, "Library");
```
Apps store their settings in `ApplicationData` and `LocalApplicationData`. Changing the locations will cause apps to no longer find their settings.

With no access to the previous location, implementing setting migration is also difficult.

Besides it being more correct, is there a tangible improvement?

Miepee replied:
> With no access to the previous location, implementing *settings migration* is also difficult.

The previous location was just `/home/.config` / `/home/.local/share` so that can be used for migration. Would need to be made clear for a release.

> is there a tangible improvement?

It follows Apple's Guidelines, which from what I know are necessary to fulfill in order to publish anything on their App Store.

tmds replied:
> It follows Apple's Guidelines, which from what I know are necessary to fulfill in order to publish anything on their App Store.

Without this change, .NET apps aren't accepted in the Apple App Store?

Miepee replied:
Not as far as I know, but am also not a macOS developer so never dealt with that in any way.
[Based on](https://developer.apple.com/app-store/review/guidelines/#before-you-submit) :
> Check whether your app follows guidance in other documentation

I'd assume so at least.

soc replied:
Is there further information on using `~/Library` for `LOCAL_APP_DATA`?

I have used `~/Library/Application Support` for that as well (https://github.com/dirs-dev/directories-jvm#basedirectories) and haven't seen issues or user complaints.

Miepee replied:
I based this mainly off [this StackOverflow post](https://stackoverflow.com/a/5084892) which in turn used this [rather old now Mac Java FAQ](https://developer.apple.com/library/archive/qa/qa1170/_index.html) and maybe the fact that the [iOS doc mentions](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW12) creating custom subdirectories.
I do however see people use `~/Library/Application Support` for [quite](https://doc.qt.io/qt-5/qstandardpaths.html) [a few](https://github.com/sindresorhus/env-paths) [other](https://github.com/adrg/xdg) [libs](https://github.com/sago007/PlatformFolders). So probably a good call to indeed change it to that instead.

soc replied:
@tmds

> backwards compatibility

Having a good migration story is important, but we can't keep pretending that accreting technical debt instead of fixing things is fine. That's one of the things that brought computing into disrepute in the wider public.

Perhaps the behavior could be predicated on the .NET version targeted, such that only code compiled for the next version gets the correct folders?

>  It will break things, and the benefit is not clear to me (besides correctness).

Writing things to `$HOME` (e. g. `~/.dotnet`) also breaks things. You cannot expect `$HOME` to be writable. (Example: I make `$HOME` read-only on every machine under my control.)

tmds replied:
For any breaking change you need to weigh the pros and cons. Calling it _technical debt_ doesn't change that.

soc replied:
Yes, and?

soc replied:
@tannergooding Agreed. If special folders are finally updated, let's make sure they are actually correct. Those who incorrectly relied on buggy behavior should make a suggestion on how to address their problems, not the people fixing things.

tmds commented on src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:
```diff
@@ -119,8 +111,23 @@ private static string GetFolderPathCoreWithoutValidation(SpecialFolder folder)
+                    }
+                    return data;
+                case SpecialFolder.MyDocuments: // same value as Personal
+                    return ReadXdgDirectory(home, "XDG_DOCUMENTS_DIR", "Documents");
```
There will be code that uses `MyDocuments` for `HOME`.
That code will break, and need to be updated to using `UserProfile` instead.

Miepee replied:
With "that code" are you referring to code inside of .NET or user made programs?

tmds replied:
Both.

marek-safar commented:
> The mobile platforms are probably the most impacted here.

This should be fine for mobile but could break VS4M when moving to net7

danmoseley commented:
Perhaps at least part is low enough concern to go into .NET 7. And anything we are less sure about can go in first build of .NET 8 giving lots of time for anyone for any widespread breaks for mobile, VS4Mac etc to show up and be considered.

eerhardt commented on src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:
```diff
@@ -393,7 +396,7 @@ public void GetSystemDirectory()
         [PlatformSpecific(TestPlatforms.AnyUnix)]  // Tests OS-specific environment
         [InlineData(Environment.SpecialFolder.UserProfile, Environment.SpecialFolderOption.None)]
-        [InlineData(Environment.SpecialFolder.MyDocuments, Environment.SpecialFolderOption.None)] // MyDocuments == Personal
+        [InlineData(Environment.SpecialFolder.MyDocuments, Environment.SpecialFolderOption.DoNotVerify)] // MyDocuments == Personal
```
Why the change to `DoNotVerify` here?

Miepee replied:
From what I remember, it was because the documents folder along with pictures/videos/music folder etc. don't exist on the test environment and thus would fail when being tested against.

eerhardt replied:
is that only for certain platforms? Which platforms does this folder not exist?

Miepee replied:
I think only for Linux.
When I made the PR initially without that change, I couldn't run the tests locally, so I relied on the CI here to check whether tests succeed or not. They at first failed due to the folders not existing, so I changed it to DoNotVerify to be consistent with the other "media" folders and make the tests pass. It has been a few months, so don't remember all the details. I force pushed a lot, so comparing the commits is difficult + the original helix logs have become invalid now.

danmoseley commented:
@eerhardt's going to sponsor this PR -- as @Miepee observed, anything about SpecialFolders tends to get stale and fail because we end up spinning on questions about breaking changes. Now we're starting a new release, we have the opportunity to take a little more risk if needed.

eerhardt commented:
I've researched the changes here and I think this change is good.

The changes outside of the `SpecialFolder.MyDocuments/Personal` changes, while breaking, are correct and straight-forward. I don't think the those changes (they are all macOS-only) are controversial.

The change of `SpecialFolder.MyDocuments/Personal` on Unix however is a pretty big breaking change. I've searched for places using this value and sent PRs where I could to fix them. The most common theme is that people were using `SpecialFolder.Personal` when they really meant `SpecialFolder.UserProfile`. Because this appears to be a common case of confusion, I opened https://github.com/dotnet/runtime/issues/75563 to hide/deprecate `Personal`. Instead, developers should explicitly pick `MyDocuments` or `UserProfile` as appropriate.

I think this change is ready to go now that net8.0 is open. We should get it in as early as possible, so we can catch any breaks this will cause. I will also write up a breaking change document for it.

@Miepee - can you rebase this change on the latest `main`? That way we can get a fresh CI run, and we can merge it when the CI passes.

Miepee committed:
Fix incorrect SpecialFolder entries for Unix

This fixes an incorrect Documents/Personal for Linux+Mac entry,
incorrect Videos Mac entry,
and an incorrect (Local) ApplicationData Mac entry.

Fix #63214

Miepee committed:
Don't make Unix tests assume Personal is Home

Miepee committed:
Use the same value for Mac's (Local)ApplicationData

Miepee commented:
Rebased! I assume the mono-llvm failure, and the tvOS cancel are unrelated?

eerhardt commented:
Failure is #75606.

I'm going to leave this open one more day for people to express any objection feedback. But if I don't see any by tomorrow, I think we should merge this change and I'll write up the breaking change doc. (Note we can always revert it later, if we get enough feedback that this was the wrong decision/change.)

Miepee commented:
No objections for 3 days now - @eerhardt wanna try merging?

soc commented:
You could check against https://github.com/dirs-dev/directories-jvm#userdirectories and its https://github.com/dirs-dev/directories-jvm/blob/main/src/main/java/dev/dirs/UserDirectories.java.

As far as I have looked, the decisions seem to match, which is a good sign.

Miepee commented:
Quickly checked MacOS, seems to match in both user dirs and base dirs.

Miepee committed:
Use NSPaths for OSX Special Folder

The Apple documentation recommends to use these instead of hardcoding the paths, so the paths that had NSPath equivalents have been replaced.

Miepee commented:
I tried applying the suggestion to use `NSSearchPath`s for OSX, but seems like there's something that the tests don't like. Is that still the failure issue from above? Because from reading the logs, I'm not getting any smarter there.

jkotas commented:
> EDIT: well, by the mention just below, i'm assuming another known infra-issue.

I do not think that this is a known issue.

I think you are hitting the unreachable assert here: https://github.com/dotnet/runtime/blob/7414af2a5f6d8d99efc27d3f5ef7a394e0b23c42/src/native/libs/System.Native/pal_searchpath.c#L10 . You need to change ifdefs around this native shim now that you have started using it in managed code.

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:
```diff
@@ -17,7 +20,7 @@ public static partial class Environment
         {
             // Get the path for the SpecialFolder
-            string path = GetFolderPathCoreWithoutValidation(folder);
+            string? path = GetFolderPathCoreWithoutValidation(folder);
             Debug.Assert(path != null);
```
This doesn't seem valid. Why would `GetFolderPathCoreWithoutValidation` return a nullable string if it will never be `null`?

Miepee replied:
`Interop.Sys.SearchPath()` returns a nullable string, which is used i.e. [here](https://github.com/dotnet/runtime/pull/68610/files#diff-a0d453801306c48ad8f563e3288251a7500e62cadee05aa27b155a6fc0864c04R100)
I also could've done something like `string? value = Interop.Sys.SearchPath(NSSearchPathDirectory.Foo); return value ?? String.Empty` for those situations, but thought this was cleaner.

eerhardt replied:
Then how is this assert valid?

Miepee replied:
...Oh I see what you mean now. I'll try fixing that.

Miepee replied:
Added a fix, lmk if it's ok.

Miepee committed:
Change System.Native CMakeLists to include OSX for SearchPath

`a17e73466ca639388a3d89f4d6be0ab9703802fc` made OSX call the native NSearchPath functions, however a stub was initially called. This fixes it to call the native functions.

Miepee commented:
Thanks [jkotas](https://github.com/jkotas) for the hint, I think I figured out the place where to do that.

Miepee committed:
Fix weird null logic for Unix GetFolderPathCore

The result from `GetFolderPathCoreWithoutValidation` should not be null for further continuation of the program, which could happen before.

eerhardt reviewed (approved):
This LGTM. Thanks @Miepee for pushing this forward.

I'll write up the breaking change doc and post it here after I merge.

RayKoopa commented:
Great work all! Once this is merged I'll begin implementing [my proposed additions to the `SpecialFolder` enum](https://github.com/dotnet/runtime/issues/70484) ASAP, as it's dependent on fixes being done here.

eerhardt commented:
Breaking change document is here: https://github.com/dotnet/docs/issues/31423

soc commented:
@eerhardt I think the line 

> `$XDG_DOCUMENTS_DIR`, if set. `$HOME/Documents`

needs to be clarified a bit, because it makes `$XDG_DOCUMENTS_DIR` look like an environment variable, not an entry from the `user-dirs.dirs` file.

In my documentation I use `XDG_DOCUMENTS_DIR` (without the dollar sign), maybe that's an option.

Miepee commented:
It is both actually.
It first checks whether the environment variable is set, and if yes, chooses that.
If the env var is not set, it reads the value from `user-dirs.dirs`
And if it can't find it there, it chooses the fallback option.

But yeah, would probably be a good idea to clarify this behaviour for user XDG directories.

soc commented:
> It is both actually.
> It first checks whether the environment variable is set, and if yes, chooses that.

Well, then that's absolutely wrong and should be corrected.

steveisok commented:
This needs corrected on Android as `GetFolderPath_Unix_PersonalIsDocumentsAndUserProfile` fails on the rolling build with:

```
Assert.Equal() Failure
↓ (pos 58)
Expected: ···tensions.Tests/files/Documents
Actual:   ···tensions.Tests/files
↑ (pos 58)
```

@Miepee some of our configurations run post PR because we have limited capacity.  It is possible to run them on PR's, but someone on the dotnet team would have to trigger them on your behalf.  Not much you can do.  Thanks for your contribution!

steveisok commented:
Cursory glance, this would need to be modified to conform with the breaking change:

https://github.com/dotnet/runtime/blob/b2af65a406bc28c25f0a25ed87b4e48c98f52fef/src/libraries/System.Private.CoreLib/src/System/Environment.Android.cs#L57-L59

Miepee commented:
I ain't that familiar with Android's file structure, should I just change `Personal` to `$HOME/Documents` and keep `LocalApplicationData` to `$HOME`?

steveisok commented:
> I ain't that familiar with Android's file structure, should I just change `Personal` to `$HOME/Documents` and keep `LocalApplicationData` to `$HOME`?

Yeah, I think that's fine.

Miepee commented:
Opened #76250 to fix that.
